File: SudokuSolver/SudokuPuzzle.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SudokuSolver
{
    public enum SudokuValue
    {
        UNASSIGNED = 0,
        ONE,
        TWO,
        THREE,
        FOUR,
        FIVE,
        SIX,
        SEVEN,
        EIGHT,
        NINE
    }

    public class SudokuPuzzle : ICloneable
    {
        private static readonly int NUM_ROWS = 9;
        private static readonly int NUM_CELLS_PER_BOX = (int)Math.Sqrt(NUM_ROWS);

        public SudokuValue[][] CurrentBoard { get; private set; }

        public SudokuPuzzle()
        {
            // initialize board
            this.ClearBoard();
        }

        public void ClearBoard()
        {
            this.CurrentBoard = new SudokuValue[NUM_ROWS][];
            for (int i = 0; i < NUM_ROWS; i++)
            {
                this.CurrentBoard[i] = new SudokuValue[NUM_ROWS];
            }
        }

        public void SetValue(int row, int col, SudokuValue value)
        {
            this.CurrentBoard[row][col] = value;
        }

        public SudokuValue GetValue(int row, int col)
        {
            return this.CurrentBoard[row][col];
        }

        public bool IsValidValue(SudokuValue sudokuValue, int row, int col)
        {
            if (this.IsValueInRow(sudokuValue, row) || this.IsValueInColumn(sudokuValue, col) || this.IsValueInBox(sudokuValue, row, col))
            {
                return false;
            }

            return true;
        }

        public bool IsSolved()
        {
            for (int i = 0; i < NUM_ROWS; i++)
            {
                for (int j = 0; j < NUM_ROWS; j++)
                {
                    var value = this.GetValue(i, j);
                    var rowCount = this.CountOccurencesInRow(value, i);
                    var colCount = this.CountOccurencesInColumn(value, j);
                    var boxCount = this.CountOccurencesInBox(value, i, j);

                    // single occurence max!
                    if (rowCount > 1 || colCount > 1 || boxCount > 1) return false;
                }
            }

            return true;
        }

        #region  [ICloneable Implementation]

        public object Clone()
        {
            var puzzle = new SudokuPuzzle();

            for (int i = 0; i < NUM_ROWS; i++)
            {
                for (int j = 0; j < NUM_ROWS; j++)
                {
                    puzzle.SetValue(i, j, this.GetValue(i, j));
                }
            }

            return puzzle;
        } 

        #endregion

        #region [Private Helper Methods]

        private bool IsValueInRow(SudokuValue value, int row)
        {
            return this.CountOccurencesInRow(value, row) > 0;
        }

        private bool IsValueInColumn(SudokuValue value, int col)
        {
            return this.CountOccurencesInColumn(value, col) > 0;
        }

        private bool IsValueInBox(SudokuValue value, int row, int col)
        {
            return this.CountOccurencesInBox(value, row, col) > 0;
        }

        private int CountOccurencesInRow(SudokuValue value, int row)
        {
            var count = 0;

            for (int i = 0; i < NUM_ROWS; i++)
            {
                if (this.CurrentBoard[row][i] == value)
                {
                    count++;
                }
            }

            return count;
        }

        private int CountOccurencesInColumn(SudokuValue value, int col)
        {
            var count = 0;

            for (int i = 0; i < NUM_ROWS; i++)
            {
                if (this.CurrentBoard[i][col] == value)
                {
                    count++;
                }
            }

            return count;
        }

        private int CountOccurencesInBox(SudokuValue value, int row, int col)
        {
            var count = 0;

            // adjust (0, 1, 2) => 0, (3, 4, 5) => 3, (6, 7, 8) => 6
            row = row - row % NUM_CELLS_PER_BOX;
            col = col - col % NUM_CELLS_PER_BOX;

            for (int i = row; i < row + NUM_CELLS_PER_BOX; i++)
            {
                for (int j = col; j < col + NUM_CELLS_PER_BOX; j++)
                {
                    if (this.CurrentBoard[i][j] == value)
                    {
                        count++;
                    }
                }
            }

            return count;
        }

        #endregion               
    }
}

File: Python/sudoku_puzzle.py
class SudokuPuzzle(object):
    NUM_ROWS = 9
    NUM_CELLS_IN_BOX = 3
    
    def __init__(self):
        self.board = []
        self.clearBoard()

    def __str__(self):
        boardString = ""
        for i in xrange(SudokuPuzzle.NUM_ROWS):
            for j in xrange(SudokuPuzzle.NUM_ROWS):
                boardString += str(self.board[i][j]) + " "

            boardString += "\n"

        return boardString[:-1]
                

    def clearBoard(self):
        self.board = []
        
        for i in xrange(SudokuPuzzle.NUM_ROWS):
            self.board.append([])
            for j in xrange(SudokuPuzzle.NUM_ROWS):
                self.board[i].append('0')

    def setValue(self, row, col, value):
        self.board[row][col] = value

    def getValue(self, row, col):
        return self.board[row][col]

    def isValidValue(self, value, row, col):
        inRow = self.isValueInRow(value, row)
        inCol = self.isValueInCol(value, col)
        inBox = self.isValueInBox(value, row, col)
        
        if inRow or inCol or inBox:
            return False

        return True
    
    def isValueInRow(self, value, row):
        return self.numOccurencesInRow(value, row) > 0

    def isValueInCol(self, value, col):
        return self.numOccurencesInCol(value, col) > 0

    def isValueInBox(self, value, row, col):
        return self.numOccurencesInBox(value, row, col) > 0

    def numOccurencesInRow(self, value, row):
        count = 0
        for i in xrange(SudokuPuzzle.NUM_ROWS):
            if self.board[row][i] == value:
                count += 1

        return count

    def numOccurencesInCol(self, value, col):
        count = 0
        for i in xrange(SudokuPuzzle.NUM_ROWS):
            if self.board[i][col] == value:
                count += 1

        return count

    def numOccurencesInBox(self, value, row, col):
        count = 0

        row = row - row % SudokuPuzzle.NUM_CELLS_IN_BOX
        col = col - col % SudokuPuzzle.NUM_CELLS_IN_BOX

        for i in xrange(row, row + SudokuPuzzle.NUM_CELLS_IN_BOX):
            for j in xrange(col, col + SudokuPuzzle.NUM_CELLS_IN_BOX):
                if self.board[i][j] == value:
                    count += 1

        return count
